File: packages/web/src/api/transactions.ts
import { apiGet, apiPost } from './client'

export interface BalanceResponse {
  memberId: string
  programId: string
  balance: number
}

export interface EarnBurnResponse {
  transactionId: string
  balance: number
  points: number
}

export function getBalance(
  tenantId: string,
  programId: string,
  memberId: string,
  idToken?: string | null
): Promise<BalanceResponse> {
  return apiGet<BalanceResponse>(
    `/api/v1/programs/${programId}/balance/${encodeURIComponent(memberId)}`,
    tenantId,
    idToken
  )
}

export function earn(
  tenantId: string,
  programId: string,
  body: { memberId: string; points: number; idempotencyKey?: string },
  idToken?: string | null
): Promise<EarnBurnResponse> {
  return apiPost<EarnBurnResponse>(
    `/api/v1/programs/${programId}/earn`,
    tenantId,
    body,
    idToken
  )
}

export function burn(
  tenantId: string,
  programId: string,
  body: { memberId: string; points: number },
  idToken?: string | null
): Promise<EarnBurnResponse> {
  return apiPost<EarnBurnResponse>(
    `/api/v1/programs/${programId}/burn`,
    tenantId,
    body,
    idToken
  )
}

File: packages/backend/src/app/routers/transactions.py
"""Transactions: earn, burn, balance, list."""

from fastapi import APIRouter, Depends, Query

from app.deps import get_tenant_id
from app.models.transactions import (
    EarnRequest,
    BurnRequest,
    BalanceResponse,
    EarnBurnResponse,
    TransactionItem,
    TransactionListResponse,
)
from app.services import transactions as svc

router = APIRouter(prefix="/programs", tags=["Transactions"])


@router.post("/{program_id}/earn", response_model=EarnBurnResponse)
def earn(program_id: str, body: EarnRequest, tenant_id: str = Depends(get_tenant_id)):
    return svc.earn(tenant_id, program_id, body.memberId, body.points, body.idempotencyKey)


@router.post("/{program_id}/burn", response_model=EarnBurnResponse)
def burn(program_id: str, body: BurnRequest, tenant_id: str = Depends(get_tenant_id)):
    return svc.burn(tenant_id, program_id, body.memberId, body.points)


@router.get("/{program_id}/balance/{member_id}", response_model=BalanceResponse)
def get_balance(program_id: str, member_id: str, tenant_id: str = Depends(get_tenant_id)):
    return svc.get_balance(tenant_id, program_id, member_id)


@router.get("/{program_id}/transactions", response_model=TransactionListResponse)
def list_transactions(
    program_id: str,
    tenant_id: str = Depends(get_tenant_id),
    memberId: str | None = Query(None, alias="memberId"),
    limit: int = Query(50, ge=1, le=100),
    nextToken: str | None = Query(None, alias="nextToken"),
):
    items, next_sk = svc.list_transactions(tenant_id, program_id, member_id=memberId, limit=limit, next_token=nextToken)
    return TransactionListResponse(
        transactions=[TransactionItem(**x) for x in items],
        nextToken=next_sk,
    )
